File: src/main/java/ru/gdlbo/search/searcher/controllers/RegistrationController.java
package ru.gdlbo.search.searcher.controllers;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestParam;
import ru.gdlbo.search.searcher.repository.*;

import java.util.Collections;

@Controller
public class RegistrationController {
    @Autowired
    private UserRepository userRepository;

    @Autowired
    private RoleRepository roleRepository;

    @Autowired
    private PasswordEncoder passwordEncoder;

    @GetMapping("/auth/register")
    public String showRegistrationForm() {
        return "registration";
    }

    @GetMapping("/auth/reg")
    public String registerUser(@RequestParam String username, @RequestParam String password) {
        if (username.isEmpty() || password.isEmpty()) {
            // Handle validation error
            System.out.println("Username or password is empty");
            return "redirect:/auth/register?error";
        }

        // Check if username already exists
        if (userRepository.findByUsername(username).isPresent()) {
            // Handle duplicate username error
            System.out.println("User already exists: " + username);
            return "redirect:/auth/register?error";
        }

        // Create a new user
        User user = new User();
        user.setUsername(username);
        user.setPassword(passwordEncoder.encode(password));

        // Ensure roles exist in the database
        Role roleAdmin = roleRepository.findByName("ROLE_ADMIN").orElseGet(() -> {
            Role newRole = new Role("ROLE_ADMIN");
            roleRepository.save(newRole);
            return newRole;
        });

        Role roleUser = roleRepository.findByName("ROLE_USER").orElseGet(() -> {
            Role newRole = new Role("ROLE_USER");
            roleRepository.save(newRole);
            return newRole;
        });

        if (username.equals("admin")) {
            user.setUserRoles(Collections.singleton(new UserRole(user, roleAdmin)));
        } else {
            user.setUserRoles(Collections.singleton(new UserRole(user, roleUser)));
        }

        // Save the user
        userRepository.save(user);

        if (userRepository.findByUsername(username).isPresent()) {
            System.out.println("User saved successfully: " + username);
            return "redirect:/auth/login";
        } else {
            System.out.println("Failed to save user: " + username);
            return "redirect:/error";
        }
    }
}

File: src/main/java/ru/gdlbo/search/searcher/controllers/web/RegistrationController.kt
package ru.gdlbo.search.searcher.controllers.web

import org.springframework.beans.factory.annotation.Autowired
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Controller
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RequestParam
import ru.gdlbo.search.searcher.repository.role.Role
import ru.gdlbo.search.searcher.repository.role.RoleRepository
import ru.gdlbo.search.searcher.repository.user.User
import ru.gdlbo.search.searcher.repository.user.UserRepository
import ru.gdlbo.search.searcher.repository.user.UserRole

@Controller
class RegistrationController {
    @Autowired
    private val userRepository: UserRepository? = null // Сервис для работы с пользователями

    @Autowired
    private val roleRepository: RoleRepository? = null // Сервис для работы с ролями пользователей

    @Autowired
    private val passwordEncoder: PasswordEncoder? = null // Сервис для шифрования паролей

    @GetMapping("/auth/register")
    fun showRegistrationForm(): String {
        return "registration"
    }

    @GetMapping("/auth/reg")
    fun registerUser(
        @RequestParam username: String,
        @RequestParam password: String
    ): String {
        if (username.isEmpty() || password.isEmpty()) {
            // Обработать ошибку валидации
            println("Имя пользователя или пароль пусты")
            return "redirect:/auth/register?error" // Перенаправление на форму регистрации с сообщением об ошибке
        }

        // Проверить, существует ли уже пользователь с таким именем
        if (userRepository!!.findByUsername(username).isPresent) {
            // Обработать ошибку занятого имени пользователя
            println("Пользователь уже существует: $username")
            return "redirect:/auth/register?error" // Перенаправление на форму регистрации с сообщением об ошибке
        }

        // Создать нового пользователя
        val user = User()
        user.username = username
        user.password = passwordEncoder!!.encode(password) // Зашифровать пароль

        // Проверить наличие ролей в базе данных
        val roleAdmin = roleRepository!!.findByName("ROLE_ADMIN").orElseGet {
            val newRole = Role("ROLE_ADMIN")
            roleRepository.save(newRole)
            newRole
        }

        val roleUser = roleRepository.findByName("ROLE_USER").orElseGet {
            val newRole = Role("ROLE_USER")
            roleRepository.save(newRole)
            newRole
        }

        // Назначение стартовой роли
        if (username == "admin") {
            user.userRoles = mutableSetOf(UserRole(user, roleAdmin))
        } else {
            user.userRoles = mutableSetOf(UserRole(user, roleUser))
        }

        // Сохранить пользователя
        userRepository.save(user)

        if (userRepository.findByUsername(username).isPresent) {
            println("Пользователь успешно создан: $username")
            return "redirect:/auth/login" // Перенаправление на страницу входа после успешной регистрации
        } else {
            println("Не удалось создать пользователя: $username")
            return "redirect:/error" // Перенаправление на страницу ошибки в случае неудачной регистрации
        }
    }
}